test/e2e/layering_test.go

```diff
@@ -0,0 +1,98 @@
+					Type: "Git",
+					Git: &buildv1.GitBuildSource{
+						URI: "https://github.com/cgwalters/fcos-derivation-example",
+						Ref: "rhcos",
+					},
+				},
+				Strategy: buildv1.BuildStrategy{
+					DockerStrategy: &buildv1.DockerBuildStrategy{},
+				},
+				Output: buildv1.BuildOutput{
+					To: &v1.ObjectReference{
+						Kind: "ImageStreamTag",
+						Name: imageStreamName + ":latest",
+					},
+				},
+			},
+		},
+	}
+	_, err = cs.BuildV1Interface.Builds(constants.MCONamespace).Create(context.TODO(), buildConfig, metav1.CreateOptions{})
+	require.Nil(t, err)
+	defer cs.BuildV1Interface.Builds(constants.MCONamespace).Delete(context.TODO(), buildName, metav1.DeleteOptions{})
+	helpers.WaitForBuild(t, cs, buildConfig.ObjectMeta.Name)
+
+	// pick a random worker node
+	unlabelFunc := helpers.LabelRandomNodeFromPool(t, cs, "worker", "node-role.kubernetes.io/infra")
+	defer unlabelFunc()
+	infraNode := helpers.GetSingleNodeByRole(t, cs, "infra")
+
+	// Run commands on the node:
+	// oc login -u kubeadmin -p <password_from_install_log> --insecure-skip-tls-verify https://api-int.<cluster_name>.<base_domain>:6443
+	// podman login -u kubeadmin -p $(oc whoami -t) image-registry.openshift-image-registry.svc:5000
+	imageRegistry := "image-registry.openshift-image-registry.svc:5000"
+	password := os.Getenv("PASSWORD")
+	clusterURL := os.Getenv("URL")
+	ocLogin := helpers.ExecCmdOnNode(t, cs, infraNode, "chroot", "/rootfs", "oc", "login", "-u", "kubeadmin", "-p", password, "--insecure-skip-tls-verify", fmt.Sprintf("https://api-int.%s:6443", clusterURL))
+	require.Contains(t, ocLogin, "Login successful.")
+	whoami := strings.TrimSuffix(helpers.ExecCmdOnNode(t, cs, infraNode, "chroot", "/rootfs", "oc", "whoami", "-t"), "\n")
+	podmanLogin := helpers.ExecCmdOnNode(t, cs, infraNode, "chroot", "/rootfs", "podman", "login", "-u", "kubeadmin", "-p", whoami, imageRegistry)
+	require.Equal(t, "Login Succeeded!\n", podmanLogin)
+
+	// rpm-ostree rebase --experimental ostree-unverified-image:docker://image-registry.openshift-image-registry.svc.cluster.local:5000/openshift-machine-config-operator/test-boot-in-cluster-image-build
+	imageURL := fmt.Sprintf("ostree-unverified-image:docker://%s/%s/%s", imageRegistry, constants.MCONamespace, imageStreamName)
+	rebase := helpers.ExecCmdOnNode(t, cs, infraNode, "chroot", "/rootfs", "rpm-ostree", "rebase", "--experimental", imageURL)
+	changesQueued := "Changes queued for next boot. Run \"systemctl reboot\" to start a reboot"
+	require.Contains(t, rebase, changesQueued)
+	// reboot
+	helpers.RebootAndWait(t, cs, infraNode)
+	// check that new image is used
+	status := helpers.ExecCmdOnNode(t, cs, infraNode, "chroot", "/rootfs", "rpm-ostree", "status")
+	require.Contains(t, status, "* "+imageURL)
+	// rollback
+	rollback := helpers.ExecCmdOnNode(t, cs, infraNode, "chroot", "/rootfs", "rpm-ostree", "rollback")
+	require.Contains(t, rollback, changesQueued)
+	helpers.RebootAndWait(t, cs, infraNode)
```

Comment:
Interesting.  Is this basically trying to return the node to a state where the existing MCD will be happy?

In these e2e tests we're running them serially against a single provisioned cluster, right?  So we probably do need this to avoid "state pollution" between tests.

But I think a bit better of a pattern we could implement would be to spin up a new disposable worker node and target that.  (This would require a bit of machineAPI glue in this repo)  Obviously not a blocker for this test, just a thought for the future.

Reply:
Agreed. I think that might interest @cheesesashimi